packages/js/src/PiwikTracker.ts

```diff
@@ -0,0 +1,209 @@
+import { TRACK_TYPES } from './constants'
+import {
+  CustomDimension,
+  TrackEventParams,
+  TrackLinkParams,
+  TrackPageViewParams,
+  TrackParams,
+  TrackSiteSearchParams,
+  UserOptions,
+} from './types'
+
+class PiwikTracker {
+  mutationObserver?: MutationObserver
+
+  constructor(userOptions: UserOptions) {
+    if (!userOptions.urlBase) {
+      throw new Error('Piwik urlBase is required.')
+    }
+    if (!userOptions.siteId) {
+      throw new Error('Piwik siteId is required.')
+    }
+
+    this.initialize(userOptions)
+  }
+
+  private initialize({
+    urlBase,
+    siteId,
+    userId,
+    trackerUrl,
+    srcUrl,
+    disabled,
+    heartBeat,
+    linkTracking = true,
+    configurations = {},
+  }: UserOptions) {
+    const normalizedUrlBase =
+      urlBase[urlBase.length - 1] !== '/' ? `${urlBase}/` : urlBase
+
+    if (typeof window === 'undefined') {
+      return
+    }
+
+    window._paq = window._paq || []
+
+    if (window._paq.length !== 0) {
+      return
+    }
+
+    if (disabled) {
+      return
+    }
+
+    this.pushInstruction(
+      'setTrackerUrl',
+      trackerUrl ?? `${normalizedUrlBase}ppms.php`,
+    )
+
+    this.pushInstruction('setSiteId', siteId)
+
+    if (userId) {
+      this.pushInstruction('setUserId', userId)
+    }
+
+    Object.entries(configurations).forEach(([name, instructions]) => {
+      if (instructions instanceof Array) {
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+        this.pushInstruction(name, ...instructions)
+      } else {
+        this.pushInstruction(name, instructions)
+      }
+    })
+
+    // accurately measure the time spent on the last pageview of a visit
+    if (!heartBeat || (heartBeat && heartBeat.active)) {
+      this.enableHeartBeatTimer((heartBeat && heartBeat.seconds) ?? 15)
+    }
+
+    // // measure outbound links and downloads
+    // // might not work accurately on SPAs because new links (dom elements) are created dynamically without a server-side page reload.
```

Comment:
```suggestion
    // measure outbound links and downloads
    // might not work accurately on SPAs because new links (dom elements) are created dynamically without a server-side page reload.
```